File: Leetcode/src/com/leetcode/math/_360_SortTransformedArray.java
package com.leetcode.math;

/**
 * _360_SortTransformedArray
 */
public class _360_SortTransformedArray {

    // 分为a > 0, a < 0, a = 0(合并为>=)
    public static int[] sortTransformedArray(int[] nums, int a, int b, int c) {
        int n = nums.length;
        int[] res = new int[n];
        int start = 0;
        int end = n - 1;
        int index = (a >= 0) ? n - 1 : 0;
        while (start <= end) {
            int startNum = getNum(nums[start], a, b, c);
            int endNum = getNum(nums[end], a, b, c);
            if (a >= 0) {
                if (startNum >= endNum) {
                    res[index--] = startNum;
                    start++;
                } else {
                    res[index--] = endNum;
                    end--;
                }
            } else {
                if (startNum >= endNum) {
                    res[index++] = endNum;
                    end--;
                } else {
                    res[index++] = startNum;
                    start++;
                }
            }
        }
        return res;
    }

    public static int getNum(int x, int a, int b, int c) {
        return a * x * x + b * x + c;
    }

    public static void main(String[] args) {
        int[] nums = {-4,-2,2,4};
        int a = 1, b = 3, c = 5;
        int[] res = sortTransformedArray(nums, a, b, c);
        for (int num : res) {
            System.out.print(num + " ");
        }

    }
}

File: Leetcode/src/com/leetcode/math/_360_Sort_Transformed_Array.py
'''
Date: 12/15/2020 10:21:24
LastEditTime: 12/15/2020 10:21:38
Description: Math, Two pointer
'''


class Solution:
    def sortTransformedArray(self, nums: List[int], a: int, b: int, c: int) -> List[int]:
        if not nums:
            return None
        start, end = 0, len(nums) - 1
        res = [0 for _ in range(len(nums))]
        index = len(nums) - 1 if a >= 0 else 0
        while start <= end:
            start_num = self.calculate(nums[start], a, b, c)
            end_num = self.calculate(nums[end], a, b, c)
            if a >= 0:
                if start_num > end_num:
                    res[index] = start_num
                    index -= 1
                    start += 1
                else:
                    res[index] = end_num
                    index -= 1
                    end -= 1
            else:
                if start_num > end_num:
                    res[index] = end_num
                    index += 1
                    end -= 1
                else:
                    res[index] = start_num
                    index += 1
                    start += 1
        return res

    def calculate(self, x, a, b, c):
        return a * x**2 + b * x + c
